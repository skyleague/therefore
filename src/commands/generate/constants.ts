import { name } from '../../../package.json'

export const generatedBy = `Generated by ${name}`
export const generatedAjv = `Generated by Ajv https://ajv.js.org/guide/managing-schemas.html#standalone-validation-code`
