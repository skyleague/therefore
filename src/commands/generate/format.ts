import { exec } from 'node:child_process'
import { promisify } from 'node:util'
import { memoize } from '@skyleague/axioms'
import type { ThereforeOutputType } from '../../lib/output/types.js'

// biome-ignore lint/suspicious/noExplicitAny: it's a third-party library
export type Prettier = any
export async function maybeLoadPrettier() {
    try {
        // @ts-ignore
        return await import('prettier')
    } catch {
        return undefined
    }
}

export async function formatContent({
    prettier,
    input,
    file,
    type,
}: {
    prettier: Prettier
    input: string
    file: string
    type: ThereforeOutputType
}): Promise<string> {
    if (prettier !== undefined) {
        const options = await prettier.resolveConfig(file)
        return prettier.format(input, {
            ...(options ?? {}),
            parser: type,
        })
    }
    return input
}

const execAsync = promisify(exec)

const getBiomeBinaryPath = memoize(async () => {
    try {
        const { stdout } = await execAsync('npx --no-install biome --version')
        if (stdout.trim()) {
            return 'npx'
        }
    } catch (error) {
        console.error('Error finding Biome:', error)
    }
    return null
})

export async function formatBiomeFiles(files: string[]) {
    const biomePath = await getBiomeBinaryPath()
    if (!biomePath) {
        console.warn('Biome not found. Skipping formatting.')
        return
    }

    const baseCommand = `${biomePath} biome check --write `
    let fileChunks: string[][]
    if (process.platform === 'win32') {
        // Windows command line length limit
        const MAX_COMMAND_LENGTH = 8180
        const remainingLength = MAX_COMMAND_LENGTH - baseCommand.length

        fileChunks = []
        let currentChunk: string[] = []
        let currentLength = 0

        for (const file of files) {
            // +1 for the space between files
            // +2 for the quotes
            const fileLength = file.length + 1 + 2

            if (currentLength + fileLength > remainingLength) {
                fileChunks.push(currentChunk)
                currentChunk = []
                currentLength = 0
            }

            currentChunk.push(`"${file}"`)
            currentLength += fileLength
        }

        if (currentChunk.length > 0) {
            fileChunks.push(currentChunk)
        }
    } else {
        // On non-Windows platforms, process all files at once
        fileChunks = [files]
    }

    try {
        const promises = fileChunks.map(async (fileChunk) => {
            const command = `${baseCommand} ${fileChunk.join(' ')}`
            const { stdout, stderr } = await execAsync(command)

            if (stderr) {
                console.warn('Biome formatting warning:', stderr)
            }
            return stdout
        })

        const results = await Promise.all(promises)
        console.log('Biome formatting completed:', results.join('\n'))
    } catch (error) {
        console.error('Error running Biome format:', error)
    }
}
