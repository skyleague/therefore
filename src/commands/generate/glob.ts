import { Nothing, fstat, isJust, isRight, normalizePath } from '@skyleague/axioms'
import fastGlob from 'fast-glob'

import path from 'node:path'

// should give the same glob behaviour as Prettier
// https://github.com/prettier/prettier/blob/main/src/cli/expand-patterns.js
// also heavily inspired by the logic there :)
export async function expandGlobs({
    patterns,
    ignore = [],
    cwd,
    extension,
}: {
    patterns: string[]
    ignore?: string[]
    cwd: string
    extension: string
}): Promise<string[]> {
    const ignoredDirectories = ['.git', '.svn', '.hg', 'node_modules']
    const globOptions = {
        dot: true,
        ignore: [...ignoredDirectories.map((dir) => `**/${dir}`), ...ignore],
    }

    const entries: string[] = (
        await Promise.allSettled(
            patterns.map(async (pattern) => {
                const absolutePath = path.resolve(cwd, pattern)
                if (ignoredDirectories.some((i) => pattern.includes(i))) {
                    return Nothing
                }

                const eitherStat = await fstat(absolutePath)
                if (isRight(eitherStat) && isJust(eitherStat.right)) {
                    if (eitherStat.right.isFile()) {
                        return fastGlob.escapePath(normalizePath(pattern))
                    }
                    if (eitherStat.right.isDirectory()) {
                        return `${fastGlob.escapePath(normalizePath(path.relative(cwd, absolutePath)))}/**/*${extension}`
                    }

                    return Nothing
                }
                if (pattern.startsWith('!')) {
                    globOptions.ignore.push(normalizePath(pattern.slice(1)))
                    return Nothing
                }
                return normalizePath(pattern)
            }),
        )
    )
        .map((result) => (result.status === 'fulfilled' ? result.value : Nothing))
        .filter(isJust)
    return await fastGlob(entries, globOptions)
}
