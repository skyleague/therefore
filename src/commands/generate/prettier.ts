import type { ThereforeOutputType } from './types.js'

import { Nothing, isNothing } from '@skyleague/axioms'

export async function maybeLoadPrettier() {
    try {
        return await import('prettier')
    } catch {
        return Nothing
    }
}

export async function formatFile(
    prettier: Awaited<ReturnType<typeof maybeLoadPrettier>>,
    input: string,
    file: string,
    type: ThereforeOutputType
): Promise<string> {
    if (isNothing(prettier)) {
        return input
    }
    const options = await prettier.resolveConfig(file)
    return prettier.format(input, {
        ...(options ?? {}),
        ...(type === 'jsonschema' ? { parser: 'json' } : {}),
    })
}
