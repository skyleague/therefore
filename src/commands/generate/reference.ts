import type { FileDefinition, TypescriptDefinition, TypescriptReference } from './types'

import type { ThereforeCst } from '../../lib/primitives/types'
import { toTypescriptDefinition } from '../../lib/visitor/typescript/typescript'

import { evaluate, sha256 } from '@skyleague/axioms'

import path from 'path'

interface ExpandedReference {
    srcPath: string
    targetPath: string
    symbolName: string
    uuid: string
    referenceName: string
    uniqueSymbolName: string
}

const localVersion = new Map<string, number>()
const referenceRegister = new Map<string, string>()

export function requireReference({
    definitions,
    current,
    ref,
    locals,
}: {
    definitions: Record<string, FileDefinition>
    current: FileDefinition
    ref: TypescriptReference
    locals: NonNullable<TypescriptDefinition['locals']>
}): ExpandedReference[] {
    const found = Object.values(definitions)
        .flatMap((d) =>
            d.symbols.map((s) => ({
                uuid: s.definition.uuid,
                symbolName: s.definition.symbolName,
                referenceName: s.definition.referenceName,
                uniqueSymbolName: s.definition.uniqueSymbolName,
                srcPath: d.srcPath,
                targetPath: d.targetPath,
            }))
        )
        .find((s) => s.uuid === ref.uuid)

    if (found === undefined) {
        // this is a locally defined variable
        let sourceSymbol: string
        const reference = evaluate(ref.reference[0])

        const refName = ref.name
        const fileHash = sha256(path.basename(current.srcPath)).slice(0, 4)
        if (refName !== undefined) {
            sourceSymbol = refName
        } else {
            if (!referenceRegister.has(reference.uuid)) {
                if (!localVersion.has(current.targetPath)) {
                    localVersion.set(current.targetPath, 0)
                }
                const localIterator = localVersion.get(current.targetPath)!
                referenceRegister.set(reference.uuid, `local${localIterator}`)
                localVersion.set(current.targetPath, localIterator + 1)
            }
            sourceSymbol = referenceRegister.get(reference.uuid)!
        }

        const { definition } = toTypescriptDefinition({
            sourceSymbol,
            schema: evaluate(reference) as ThereforeCst,
            fileHash,
            exportSymbol: ref.exportSymbol,
            locals,
        })
        current.symbols.push({
            uuid: ref.uuid,
            definition,
            symbolName: sourceSymbol,
            typeOnly: true,
        })

        // make the DFS well defined
        const extra = definition.references.map((r) => requireReference({ definitions, current, ref: r, locals })).flat()
        return [
            ...extra,
            {
                uuid: ref.uuid,
                symbolName: definition.symbolName,
                srcPath: current.srcPath,
                targetPath: current.targetPath,
                referenceName: definition.referenceName,
                uniqueSymbolName: definition.uniqueSymbolName,
            },
        ]
    }
    return [
        {
            uuid: ref.uuid,
            symbolName: found.symbolName,
            srcPath: found.srcPath,
            targetPath: found.targetPath,
            referenceName: found.referenceName,
            uniqueSymbolName: found.uniqueSymbolName,
        },
    ]
}
