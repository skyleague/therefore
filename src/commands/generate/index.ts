import { generate } from './generate'

import { getExtension } from '../../common/template/path'

import type { Argv } from 'yargs'

export function builder(yargs: Argv) {
    return yargs
        .option('files', {
            alias: 'f',
            demandOption: true,
            describe: 'globs to scan for schemas',
            type: 'array',
        })
        .option('ignore-pattern', {
            alias: 'i',
            demandOption: true,
            default: ['**/*.d.ts', 'node_modules'],
            describe: 'globs to exclude',
            type: 'array',
        })
        .option('compile', {
            default: true,
            type: 'boolean',
        })
        .option('ext', {
            default: '.schema.ts',
            type: 'string',
        })
        .option('out-ext', {
            default: '.type.ts',
            type: 'string',
        })
        .option('clean', {
            default: false,
            type: 'boolean',
        })
}

export async function handler(argv: ReturnType<typeof builder>['argv']): Promise<void> {
    const { files = [], 'ignore-pattern': ignorePatterns, ext, 'out-ext': outExt, compile, clean } = await argv

    await generate({
        globs: files.map((f) => f.toString()),
        ignore: ignorePatterns.map((i) => i.toString()),
        extension: ext,
        outputFileRename: (path: string) => path.replace(getExtension(path)!, outExt),
        compile,
        clean,
    })
}

export default {
    command: ['generate', '*'],
    describe: 'generate therefore types',
    builder,
    handler,
}
