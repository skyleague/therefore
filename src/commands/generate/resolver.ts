import { generatedBy } from './constants.js'
import { requireReference } from './reference.js'
import type { FileDefinition, OutputFile, ReferenceData, TypescriptDefinition } from './types.js'

import packageJson from '../../../package.json' assert { type: 'json' }
import { writeThereforeSchema } from '../../lib/visitor/typescript/typescript.js'
import { createWriter } from '../../lib/writer.js'

import { evaluate, hasPropertiesDefined, unique } from '@skyleague/axioms'

import path from 'node:path'

const { version } = packageJson

export function renderTypescriptSchemaHeader(definition: FileDefinition) {
    const writer = createWriter()
    writer
        .writeLine('/**')
        .writeLine(`* ${generatedBy}@v${version}`)
        .writeLine('* Do not manually touch this')
        .newLineIfLastNot()
        .closeComment()
        .writeLine('/* eslint-disable */')

    for (const importFile of definition.symbols
        .filter(hasPropertiesDefined(['schemaFile']))
        .filter((i) => !i.typeOnly && i.compiledFile === undefined)
        .map((i) => `import ${i.symbolName}Schema from '${i.schemaFile}' assert { type: "json" }`)
        .sort()) {
        writer.writeLine(importFile)
    }

    for (const importFile of [...new Set(definition.symbols.map((i) => i.definition.imports).flat())].sort()) {
        writer.writeLine(importFile)
    }

    for (const importFile of Object.entries(definition.dependencies)
        .map(([targetPath, deps]) => ({
            targetPath,
            deps: `{ ${[...new Set(deps)]
                .sort()
                .map((d) => (definition.dependencyUsesValue[targetPath]?.has(d) ? d : `type ${d}`))
                .join(', ')} }`,
        }))
        .filter(({ targetPath }) => targetPath !== definition.targetPath)
        .map(({ targetPath: f, deps }) => {
            const otherPath = path.relative(path.dirname(definition.targetPath), f).replace('.ts', '').replace(/\\/g, '/')
            return `import ${deps} from '${otherPath.startsWith('.') ? otherPath : `./${otherPath}`}.js'`
        })
        .sort()) {
        writer.writeLine(importFile)
    }

    writer.newLine().newLine()

    return writer.toString()
}

export function renderTypescriptSchemaContent(definition: FileDefinition) {
    const writer = createWriter()
    for (const symbol of Object.values(definition.symbols)) {
        writer.writeLine(symbol.definition.declaration)
        if (symbol.definition.isExported && !symbol.typeOnly) {
            writer.writeLine(
                writeThereforeSchema({
                    uuid: symbol.definition.uuid,
                    schemaReference: `${symbol.symbolName}Schema`,
                    validatorFile: symbol.compiledFile,
                    description: evaluate(symbol.definition.schema).description,
                })
            )
        }
    }

    return writer.toString()
}

export function resolveTypescriptSchema({
    definitions,
    targetPath,
    cwd,
    outputFiles,
    localReferences,
}: {
    definitions: Record<string, FileDefinition>
    targetPath: string
    cwd: string
    outputFiles: OutputFile[]
    localReferences: Record<string, ReferenceData>
}) {
    const definition = definitions[targetPath]!
    const locals: NonNullable<TypescriptDefinition['locals']> = {}
    const required = definition.symbols
        .flatMap((i) =>
            i.definition.references.map((r) => requireReference({ definitions, current: definition, ref: r, locals }))
        )
        .flat(2)
    const uniqueRequires = [...unique(required, (a, b) => a.uuid === b.uuid)]

    const templateContent = renderTypescriptSchemaContent(definition)
    for (const r of uniqueRequires) {
        definition.dependencies[r.targetPath] ??= []
        definition.dependencies[r.targetPath]?.push(r.symbolName)

        if (templateContent.includes(`${r.uuid}:symbolName~value`)) {
            definition.dependencyUsesValue[r.targetPath] ??= new Set()
            definition.dependencyUsesValue[r.targetPath]?.add(r.symbolName)
        }
    }

    for (const symbol of definition.symbols) {
        localReferences[symbol.definition.uuid] ??= {
            symbolName: symbol.definition.symbolName,
            referenceName: symbol.definition.referenceName,
            uniqueSymbolName: symbol.definition.uniqueSymbolName,
            srcPath: targetPath,
        }
    }

    const references: Record<string, ReferenceData> = Object.fromEntries([
        ...required.map(
            (r) =>
                [
                    r.uuid,
                    {
                        symbolName: r.symbolName,
                        uniqueSymbolName: r.uniqueSymbolName,
                        referenceName: r.referenceName,
                        srcPath: r.targetPath,
                    },
                ] as const
        ),
    ])

    const schemaFile = path.relative(cwd, definition.targetPath).replace(/\\/g, '/')
    const relativeSource = `./${path.relative(path.dirname(schemaFile), definition.srcPath).replace(/\.ts$/g, '')}`
    outputFiles.push(
        {
            targetPath,
            relativeSource,
            type: 'typescript',
            template: `${renderTypescriptSchemaHeader(definition)}${templateContent}`,
            data: { ...localReferences, ...references },
            prettify: true,
        },
        ...definition.attachedFiles.map((j) => ({
            targetPath: path.relative(cwd, j.targetPath),
            relativeSource,
            type: j.type,
            template: j.content,
            data: { ...localReferences, ...references },
            prettify: j.prettify,
            clean: j.clean,
        }))
    )
}
