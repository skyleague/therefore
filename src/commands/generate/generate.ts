import { generatedAjv, generatedBy } from './constants.js'
import { expandGlobs } from './glob.js'
import { formatFile, maybeLoadPrettier } from './prettier.js'
import { resolveTypescriptSchema } from './resolver.js'
import type { FileDefinition, OutputFile, ReferenceData, ThereforeOutputType } from './types.js'

import { getExtension, replaceExtension } from '../../common/template/path.js'
import { renderTemplate } from '../../common/template/template.js'
import type { ThereforeNode } from '../../lib/cst/cst.js'
import type { ThereforeCst } from '../../lib/primitives/types.js'
import { isThereforeExport } from '../../lib/primitives/types.js'
import { toJsonSchema } from '../../lib/visitor/jsonschema/jsonschema.js'
import { prepass } from '../../lib/visitor/prepass/prepass.js'
import { toTypescriptDefinition } from '../../lib/visitor/typescript/typescript.js'

import { entriesOf, enumerate, groupBy, isDefined, range, second } from '@skyleague/axioms'
import decamelize from 'decamelize'

import fs from 'node:fs'
import path from 'node:path'
import { fileURLToPath } from 'node:url'

const __dirname = path.dirname(fileURLToPath(import.meta.url))
async function requireModule(module: string): Promise<Record<string, ThereforeNode | unknown>> {
    const relative = path.relative(__dirname, module).replace(/\\/g, '/').replace(/\.ts$/, '.js')

    const mod = (await import(relative.startsWith('.') ? relative : `./${relative}`)) as Record<string, unknown>
    return (mod.default ?? mod) as Record<string, ThereforeNode | unknown>
}

export async function loadSymbol({
    srcPath,
    basePath,
    symbol,
    sourceSymbol,
    compile,
    definitions,
    entry,
    fileSuffix,
    filePath,
    outputFileRename,
    appendSourceSymbolName: appendSourceName = false,
    rootPath = srcPath,
}: {
    srcPath: string
    basePath: string
    symbol: ThereforeCst
    sourceSymbol: string
    compile: boolean
    definitions: Record<string, FileDefinition | undefined>
    entry: string
    fileSuffix?: string | undefined
    filePath?: string | undefined
    outputFileRename: (path: string) => string
    appendSourceSymbolName?: boolean
    rootPath?: string
}) {
    const simplified = prepass(symbol)
    let targetDir = '.'
    if (simplified.type === 'custom') {
        if (simplified.value.fileSuffix !== undefined) {
            fileSuffix = simplified.value.fileSuffix
        }
        if (simplified.value.filePath !== undefined) {
            const dir = path.dirname(simplified.value.filePath)
            const basename = path.basename(simplified.value.filePath)
            filePath = basename
            targetDir = dir
        }
    }

    const { base: baseName, dir: baseDir } = path.parse(srcPath)
    const ext = getExtension(srcPath) ?? ''
    const targetBaseDir = path.relative(basePath, path.join(baseDir, targetDir))
    const targetSrcPath = `${targetBaseDir}/${(filePath ?? baseName).replace(ext, '')}${ext}`
    const targetPath = fileSuffix !== undefined ? replaceExtension(targetSrcPath, fileSuffix) : outputFileRename(targetSrcPath)

    definitions[targetPath] ??= {
        srcPath: path.relative(basePath, rootPath),
        targetPath,
        attachedFiles: [],
        symbols: [],
        dependencies: {},
    }

    const file = definitions[targetPath]!
    // check if symbol is not seen yet
    if (file.symbols.find((s) => s.uuid === simplified.uuid)) {
        return
    }
    const symbolName = simplified.name ?? sourceSymbol

    sourceSymbol = appendSourceName && simplified.name !== undefined ? `${sourceSymbol}.${simplified.name}` : symbolName
    const schemaName = decamelize(symbolName, { separator: '-' })
    const schemaFile = `./schemas/${schemaName}.schema.json`
    const compiledFile = `./schemas/${schemaName}.schema.js`

    const nodeIsCompiled = simplified.description.validator?.compile ?? compile

    const jsonschema = await toJsonSchema(simplified, nodeIsCompiled)
    const { definition, subtrees } = toTypescriptDefinition({ sourceSymbol, symbolName, schema: simplified })

    file.symbols.push(
        ...Object.values(definition.locals ?? {})
            .filter(isDefined)
            .map((local) => ({
                uuid: local.uuid,
                symbolName: local.symbolName,
                definition: local,
                typeOnly: true,
            }))
    )

    file.symbols.push({
        uuid: simplified.uuid,
        symbolName,
        definition: definition,
        schemaFile,
        compiledFile: nodeIsCompiled ? compiledFile : undefined,
        typeOnly: simplified.description.validator?.enabled !== true,
    })

    if (definition.isExported && simplified.description.validator?.enabled) {
        const cleanSchemasFolder = () => {
            const targetFolder = path.dirname(path.join(targetBaseDir, compiledFile))
            if (fs.existsSync(targetFolder)) {
                console.warn(`Cleaning ${targetFolder}`)
                fs.rmSync(targetFolder, { force: true, recursive: true })
            }
        }
        if (jsonschema.compiled) {
            file.attachedFiles.push({
                targetPath: path.join(targetBaseDir, compiledFile),
                content: `/**\n * ${generatedAjv} \n * eslint-disable\n */\n${jsonschema.code}`,
                prettify: false,
                type: 'validator',
                clean: cleanSchemasFolder,
            })
        } else {
            file.attachedFiles.push({
                targetPath: path.join(targetBaseDir, schemaFile),
                content: JSON.stringify(jsonschema.schema, null, 2),
                prettify: true,
                type: 'jsonschema',
                clean: cleanSchemasFolder,
            })
        }

        console.debug(` - found ${definition.symbolName}`)
    }
    for (const { node: subSymbol, fileSuffix: subFileSuffix, filePath: subFilePath } of subtrees) {
        await loadSymbol({
            symbol: subSymbol,
            basePath,
            sourceSymbol,
            compile,
            definitions,
            srcPath: path.join(basePath, targetPath),
            rootPath: srcPath,
            fileSuffix: subFileSuffix,
            filePath: subFilePath,
            entry,
            outputFileRename,
            appendSourceSymbolName: true,
        })
    }
}

export async function scanModule({
    entry,
    srcPath,
    basePath,
    compile,
    definitions,
    require = requireModule,
    outputFileRename,
}: {
    entry: string
    srcPath: string
    basePath: string
    compile: boolean
    definitions: Record<string, FileDefinition>
    require?: (module: string) => Promise<Record<string, ThereforeNode | unknown>>
    outputFileRename: (path: string) => string
}) {
    const module = await require(entry)

    for (const [sourceSymbol, symbolPromise] of Object.entries(module)) {
        const symbol = await symbolPromise

        if (!isThereforeExport(symbol)) {
            continue
        }

        await loadSymbol({ symbol, sourceSymbol, compile, definitions, srcPath, entry, basePath, outputFileRename })
    }
}

export async function scanFiles({
    files,
    basePath,
    compile,
    outputFileRename,
}: {
    files: string[]
    basePath: string
    compile: boolean
    outputFileRename: (path: string) => string
}): Promise<Record<string, FileDefinition>> {
    const definitions: Record<string, FileDefinition> = {}
    for (const entry of files) {
        const srcPath = path.resolve(basePath, entry)
        let isGenerated = false
        const strippedPath = srcPath.substr(0, srcPath.lastIndexOf('.'))
        for (const ext of ['.ts', '.js']) {
            const p = `${strippedPath}${ext}`
            isGenerated ||= fs.existsSync(p) && fs.readFileSync(p).includes(generatedBy)
        }
        if (isGenerated) {
            console.debug(`scanning ${entry}`)
            console.debug(` * skipping generated schema`)
            continue
        }

        console.log(`scanning ${entry}`)
        try {
            await scanModule({ entry, basePath, srcPath, compile, definitions, outputFileRename })
        } catch (e: unknown) {
            const error = e as Error
            console.debug(error.message, error.stack)
            throw error
        }
    }
    return definitions
}

export async function compileOutputFiles(
    entries: string[],
    { compile = false, cwd, outputFileRename }: { cwd: string; compile: boolean; outputFileRename: (path: string) => string }
): Promise<
    {
        targetPath: string
        template: string
        data: Record<string, ReferenceData>
        type: ThereforeOutputType
        prettify: boolean
        clean?: () => void
    }[]
> {
    const outputFiles: OutputFile[] = []
    const definitions = await scanFiles({ files: entries, basePath: cwd, compile, outputFileRename })

    const localReferences = {}

    for (const targetPath of Object.keys(definitions)) {
        resolveTypescriptSchema({ definitions, targetPath, cwd, outputFiles, localReferences })
    }

    return compile ? outputFiles : outputFiles.filter((f) => f.type !== 'validator')
}

export async function generate({
    globs,
    ignore,
    extension,
    compile,
    outputFileRename,
    clean,
}: {
    globs: string[]
    ignore: string[]
    extension: string
    compile: boolean
    outputFileRename: (path: string) => string
    clean: boolean
}): Promise<void> {
    const cwd = process.cwd()
    const entries = await expandGlobs({ patterns: globs, ignore: [`!${extension}`, ...ignore], cwd, extension })

    const outputFiles = await compileOutputFiles(entries, {
        cwd,
        outputFileRename,
        compile,
    })

    if (clean) {
        for (const of of outputFiles) {
            if (of.clean !== undefined) {
                of.clean()
            }
        }
    }

    const prettier = maybeLoadPrettier()

    const data = outputFiles.reduce<Record<string, ReferenceData>>(
        (agg, x) => ({
            ...agg,
            ...x.data,
        }),
        {}
    )

    const duplicates = entriesOf(groupBy(entriesOf(data), ([, values]) => `${values.srcPath}${values.symbolName}`)).filter(
        ([, records]) => records.length > 1
    )
    for (const duplicateGroup of duplicates.map((d) => second(d))) {
        for (const [i, [key]] of enumerate(duplicateGroup)) {
            // data[key].referenceName = `${data[key].referenceName}${i}`
            data[key]!.symbolName = `${data[key]!.symbolName}${i}`
            data[key]!.uniqueSymbolName = `${data[key]!.uniqueSymbolName}${i}`
        }
    }

    const variablesMap = Object.fromEntries(
        Object.entries(data)
            .map(([key, vals]) => Object.entries(vals).map(([varName, value]) => [`${key}:${varName}`, value] as const))
            .flat()
    )

    for (const { targetPath, template, type, prettify } of outputFiles) {
        fs.mkdirSync(path.dirname(targetPath), { recursive: true })
        // allow nested references up to 3 deep
        let contents = template
        for (const _ of range(3)) {
            contents = renderTemplate(contents, variablesMap)
        }
        fs.writeFileSync(targetPath, prettify ? await formatFile(prettier, contents, targetPath, type) : contents)
    }
}
