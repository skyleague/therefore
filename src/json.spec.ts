import { compileOutputFiles } from './commands/generate/generate.js'

import { expect, it } from 'vitest'

it('json', async () => {
    expect(
        await compileOutputFiles(['examples/json/json.schema.ts'], {
            outputFileRename: (file: string) => file.replace('.ts', '.type.ts'),
            cwd: process.cwd(),
            compile: false,
        })
    ).toMatchSnapshot()
})

it('simple', async () => {
    expect(
        await compileOutputFiles(['examples/json/simple.schema.ts'], {
            outputFileRename: (file: string) => file.replace('.ts', '.type.ts'),
            cwd: process.cwd(),
            compile: false,
        })
    ).toMatchSnapshot()
})
