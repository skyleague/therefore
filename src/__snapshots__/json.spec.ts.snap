// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`json 1`] = `
[
  {
    "data": {
      "00010-000": {
        "referenceName": "{{00010-000:symbolName}}",
        "srcPath": "examples/json/json.schema.type.ts",
        "symbolName": "Json",
        "uniqueSymbolName": "{{00010-000:symbolName}}",
      },
      "00012-000": {
        "referenceName": "{{00012-000:symbolName}}",
        "srcPath": "examples/json/json.schema.type.ts",
        "symbolName": "JsonAdv",
        "uniqueSymbolName": "{{00012-000:symbolName}}",
      },
      "00021-000": {
        "referenceName": "{{00021-000:symbolName}}",
        "srcPath": "examples/json/json.schema.type.ts",
        "symbolName": "JsonLocal",
        "uniqueSymbolName": "{{00021-000:symbolName}}4e98",
      },
    },
    "prettify": true,
    "relativeSource": "./json.schema",
    "targetPath": "examples/json/json.schema.type.ts",
    "template": "/**
* Generated by @skyleague/therefore@v1.0.0-local
* Do not manually touch this
*/
/* eslint-disable */
import jsonAdvSchema from './schemas/json-adv.schema.json'
import jsonSchema from './schemas/json.schema.json'
import AjvValidator from 'ajv'
import type { ValidateFunction } from 'ajv'


export type {{00010-000:symbolName}} = string | null | boolean | number | {
    [k: string]: ( {{00010-000:referenceName}} ) | undefined
} | ({{00010-000:referenceName}})[]

export const {{00010-000:symbolName}} = {
    validate: new AjvValidator({"strict":true,"strictTypes":true,"useDefaults":true,"removeAdditional":false,"strictSchema":false,"code":{"optimize":true}}).compile<{{00010-000:symbolName}}>(jsonSchema),
    schema: jsonSchema,
    is: (o: unknown): o is {{00010-000:symbolName}} => {{00010-000:symbolName}}.validate(o) === true,
} as const

export type {{00012-000:symbolName}} = {{00021-000:referenceName}}

export const {{00012-000:symbolName}} = {
    validate: new AjvValidator({"strict":true,"strictTypes":true,"useDefaults":true,"removeAdditional":false,"strictSchema":false,"code":{"optimize":true}}).compile<{{00012-000:symbolName}}>(jsonAdvSchema),
    schema: jsonAdvSchema,
    is: (o: unknown): o is {{00012-000:symbolName}} => {{00012-000:symbolName}}.validate(o) === true,
    assert: (o: unknown) => {
        if (!{{00012-000:symbolName}}.validate(o)) {
            throw new AjvValidator.ValidationError({{00012-000:symbolName}}.validate.errors ?? [])
        }
    },
} as const

type {{00021-000:symbolName}} = string | null | boolean | number | {
    [k: string]: ( {{00012-000:referenceName}} ) | undefined
} | ({{00012-000:referenceName}})[]

",
    "type": "typescript",
  },
  {
    "data": {
      "00010-000": {
        "referenceName": "{{00010-000:symbolName}}",
        "srcPath": "examples/json/json.schema.type.ts",
        "symbolName": "Json",
        "uniqueSymbolName": "{{00010-000:symbolName}}",
      },
      "00012-000": {
        "referenceName": "{{00012-000:symbolName}}",
        "srcPath": "examples/json/json.schema.type.ts",
        "symbolName": "JsonAdv",
        "uniqueSymbolName": "{{00012-000:symbolName}}",
      },
      "00021-000": {
        "referenceName": "{{00021-000:symbolName}}",
        "srcPath": "examples/json/json.schema.type.ts",
        "symbolName": "JsonLocal",
        "uniqueSymbolName": "{{00021-000:symbolName}}4e98",
      },
    },
    "prettify": true,
    "relativeSource": "./json.schema",
    "targetPath": "examples/json/schemas/json.schema.json",
    "template": "{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "anyOf": [
    {
      "type": "string"
    },
    {
      "type": "null"
    },
    {
      "type": "boolean"
    },
    {
      "type": "number"
    },
    {
      "type": "object",
      "additionalProperties": {
        "$ref": "#"
      }
    },
    {
      "type": "array",
      "items": {
        "$ref": "#"
      }
    }
  ]
}",
    "type": "jsonschema",
  },
  {
    "data": {
      "00010-000": {
        "referenceName": "{{00010-000:symbolName}}",
        "srcPath": "examples/json/json.schema.type.ts",
        "symbolName": "Json",
        "uniqueSymbolName": "{{00010-000:symbolName}}",
      },
      "00012-000": {
        "referenceName": "{{00012-000:symbolName}}",
        "srcPath": "examples/json/json.schema.type.ts",
        "symbolName": "JsonAdv",
        "uniqueSymbolName": "{{00012-000:symbolName}}",
      },
      "00021-000": {
        "referenceName": "{{00021-000:symbolName}}",
        "srcPath": "examples/json/json.schema.type.ts",
        "symbolName": "JsonLocal",
        "uniqueSymbolName": "{{00021-000:symbolName}}4e98",
      },
    },
    "prettify": true,
    "relativeSource": "./json.schema",
    "targetPath": "examples/json/schemas/json-adv.schema.json",
    "template": "{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "$ref": "#/$defs/{{00021-000:uniqueSymbolName}}",
  "$defs": {
    "{{00021-000:uniqueSymbolName}}": {
      "title": "jsonLocal",
      "anyOf": [
        {
          "type": "string"
        },
        {
          "type": "null"
        },
        {
          "type": "boolean"
        },
        {
          "type": "number"
        },
        {
          "type": "object",
          "additionalProperties": {
            "$ref": "#"
          }
        },
        {
          "type": "array",
          "items": {
            "$ref": "#"
          }
        }
      ]
    }
  }
}",
    "type": "jsonschema",
  },
]
`;

exports[`simple 1`] = `
[
  {
    "data": {
      "0005-000": {
        "referenceName": "{{0005-000:symbolName}}",
        "srcPath": "examples/json/simple.schema.type.ts",
        "symbolName": "Simple",
        "uniqueSymbolName": "{{0005-000:symbolName}}",
      },
    },
    "prettify": true,
    "relativeSource": "./simple.schema",
    "targetPath": "examples/json/simple.schema.type.ts",
    "template": "/**
* Generated by @skyleague/therefore@v1.0.0-local
* Do not manually touch this
*/
/* eslint-disable */
import simpleSchema from './schemas/simple.schema.json'
import AjvValidator from 'ajv'
import type { ValidateFunction } from 'ajv'


export type {{0005-000:symbolName}} = number | ({{0005-000:referenceName}})[]

export const {{0005-000:symbolName}} = {
    validate: new AjvValidator({"strict":true,"strictTypes":true,"useDefaults":true,"removeAdditional":false,"strictSchema":false,"code":{"optimize":true}}).compile<{{0005-000:symbolName}}>(simpleSchema),
    schema: simpleSchema,
    is: (o: unknown): o is {{0005-000:symbolName}} => {{0005-000:symbolName}}.validate(o) === true,
    assert: (o: unknown) => {
        if (!{{0005-000:symbolName}}.validate(o)) {
            throw new AjvValidator.ValidationError({{0005-000:symbolName}}.validate.errors ?? [])
        }
    },
} as const

",
    "type": "typescript",
  },
  {
    "data": {
      "0005-000": {
        "referenceName": "{{0005-000:symbolName}}",
        "srcPath": "examples/json/simple.schema.type.ts",
        "symbolName": "Simple",
        "uniqueSymbolName": "{{0005-000:symbolName}}",
      },
    },
    "prettify": true,
    "relativeSource": "./simple.schema",
    "targetPath": "examples/json/schemas/simple.schema.json",
    "template": "{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "anyOf": [
    {
      "type": "number"
    },
    {
      "type": "array",
      "items": {
        "$ref": "#"
      }
    }
  ]
}",
    "type": "jsonschema",
  },
]
`;
